language_models: Add tool use support for Mistral models

imumesh18 opened the pull request:
Closes https://github.com/zed-industries/zed/issues/29855

Implement tool use handling in Mistral provider, including mapping tool call events and updating request construction. Add support for tool_choice and parallel_tool_calls in Mistral API requests.

This works fine with all the existing models. Didn't touched anything else but for future. Fetching models using their models api, deducting tool call support, parallel tool calls etc should be done from model data from api response.

<img width="547" alt="Screenshot 2025-05-06 at 4 52 37 PM" src="https://github.com/user-attachments/assets/4c08b544-1174-40cc-a40d-522989953448" />

Tasks:

- [x] Add tool call support
- [x] Auto Fetch models using mistral api
- [x] Add tests for mistral crates.
- [x] Fix mistral configurations for llm providers.

Release Notes:

- agent: Add tool call support for existing mistral models

imumesh18 committed:
Add tool use support for Mistral models

Implement tool use handling in Mistral provider, including mapping
tool call events and updating request construction. Add support for
tool_choice and parallel_tool_calls in Mistral API requests.

notpeter committed:
Merge branch 'main' into mistral-tool-calls

imumesh18 committed:
Add Mistral provider support to assistant settings

notpeter commented:
I was able to test this, it worked some, but I was able to trigger the following error:
> Failed to connect to Mistral API: 400 Bad Request {"object":"error","message":"Unexpected role 'user' after role 'tool'","type":"invalid_request_error","param":null,"code":null}

<img width="756" alt="Screenshot 2025-05-06 at 14 00 53" src="https://github.com/user-attachments/assets/9ea71b44-ad4e-4b2f-b2b9-f7d151a86c1e" />

[mistral-tool.md](https://github.com/user-attachments/files/20068408/mistral-tool.md)

imumesh18 commented:
Hey @notpeter thanks for testing, can you help me with the which model did you try this with. I am currently unable to reproduce this at the mode. Will see if I am able to do it with the specific model.

notpeter commented:
It was with `codestral-latest`.  I haven't been able to come up with a consistent repro. :(

imumesh18 committed:
Merge branch 'main' into mistral-tool-calls

imumesh18 committed:
fix

imumesh18 committed:
feat: add log dependency and implement model fetching functionality

imumesh18 committed:
📎

imumesh18 committed:
Fix models endpoint path in fetch_models function

imumesh18 committed:
Merge branch 'main' into mistral-tool-calls

imumesh18 committed:
Handle image tool results in Mistral message conversion

imumesh18 committed:
add tests

vlebert commented:
@notpeter I think the last mistral-medium should be a lot more reliable. I have very good results with other agentic tools such as Cline or Aider.dev

bennetbo commented on crates/language_models/src/provider/mistral.rs:
```diff
@@ -131,18 +142,58 @@ impl State {
             }),
         });
 
+        if let Ok(env_api_key) = std::env::var(MISTRAL_API_KEY_VAR) {
```
Why do we need this?

imumesh18 replied:
My bad, that environment variable check was a leftover from debugging the automatic model loading feature. It's no longer needed, we can remove it.

imumesh18 replied:
Removed it. Thanks for pointing it out.

bennetbo committed:
Cleanup

bennetbo committed:
Merge branch 'main' into mistral-tool-calls

bennetbo committed:
Fix `supports_tool_choice`

imumesh18 committed:
cleanup code

bennetbo commented:
One thing that i do not love is that fetch models seems to return a very long list for available models (150 for me). It does not seem like most people are interested in using most of them, so i think it just adds a lot of unnecessary noise.
Is there any way that we can de-dup these? Or go back to the original approach and not fetch them at all?

imumesh18 commented:
Hey @bennetbo got it. I guess in my guess it was less so I didn't see this issue. In that case i can revert model fetch and have a hardcoded models in this pr. Probably we can discuss a better approach for exposing all models as a part of seperate pr wdyt?

bennetbo commented:
Sounds good 👍🏻

vlebert commented:
Maybi only `-latest` models ?

imumesh18 committed:
remove model auto fetch

imumesh18 committed:
fix

imumesh18 committed:
add supports_tools in the available model settings

imumesh18 commented:
Hey @bennetbo , I have removed the code changes for fetching the model automatically. Also I have added the `language_models.mistral.available_models[n].supports_tools` so that users can define their own models and use it with tools. I have attached the demo for custom model usage. It's ready for review now and let me know if there is anyother changes you want me to do.

https://github.com/user-attachments/assets/a7de2687-f9bc-4a1e-8284-f402944e9b9d

imumesh18 committed:
fix

imumesh18 committed:
Merge branch 'main' into mistral-tool-calls

imumesh18 committed:
Merge branch 'main' into mistral-tool-calls

notpeter committed:
Pull in docs from:
- https://github.com/zed-industries/zed/pull/30755

notpeter committed:
Lol. linebreak

imumesh18 committed:
Merge branch 'main' into mistral-tool-calls

KhazAkar commented:
Tested it today myself, but it needs 'reminding' to use tools, be it `mistral-small-latest` or `codestral-latest`.

vlebert commented:
Current best mistral agentic model is mistral medium latest actually

bennetbo reviewed (approved):
Thanks!
